File: LeetCode 207- course schedule/course_schedule.go
package main

import "fmt"

func canFinish(numCourses int, prerequisites [][]int) bool {
	graph := buildGraph(numCourses, prerequisites)
	fmt.Printf("adjacency list: %+v\n\n", graph)

	visiting := map[int]bool{}
	visited := map[int]bool{}

	for node := 0; node < numCourses; node += 1 {
		fmt.Println("------------------------------------------")
		fmt.Println("dfs node ", node)
		if detectCycle(graph, node, visiting, visited) {
			return false
		}
	}

	return true
}

func detectCycle(
	graph map[int][]int,
	node int,
	visiting, visited map[int]bool) bool {
	if _, found := visited[node]; found {
		fmt.Printf("node %d is already visited(black) -> skip\n", node)
		return false
	}

	if _, found := visiting[node]; found {
		fmt.Printf("node %d is in visiting(gray) -> a cycle is detected\n\n", node)
		return true
	}

	visiting[node] = true
	fmt.Printf("nodes in visiting(gray): %+v\n", visiting)
	fmt.Printf("nodes in visited(black): %+v\n\n", visited)
	for _, descendant := range graph[node] {
		fmt.Printf("current node: node %d\n", node)
		fmt.Printf("visit descendant: node %d\n", descendant)
		if detectCycle(graph, descendant, visiting, visited) {
			return true
		}
	}

	delete(visiting, node)
	visited[node] = true
	fmt.Printf("finish explore node %d\n", node)
	fmt.Printf("nodes in visiting(gray): %+v\n", visiting)
	fmt.Printf("nodes in visited(black): %+v\n\n", visited)
	return false
}

func buildGraph(numNodes int, prerequisites [][]int) map[int][]int {
	graph := map[int][]int{}
	for i := 0; i < numNodes; i++ {
		graph[i] = []int{}
	}

	for _, prerequisite := range prerequisites {
		a, b := prerequisite[0], prerequisite[1]
		graph[b] = append(graph[b], a)
	}

	return graph
}

func main() {
	numCourses := 6
	prerequisites := [][]int{
		{1, 0},
		{3, 0},
		{5, 0},
		{2, 1},
		{3, 1},
		{4, 3},
		{5, 4},
		{3, 5},
	}
	fmt.Println("can finish courses?", canFinish(numCourses, prerequisites))
}

/*
output:
adjacency list: map[0:[1 3 5] 1:[2 3] 2:[] 3:[4] 4:[5] 5:[3]]

------------------------------------------
dfs node  0
nodes in visiting(gray): map[0:true]
nodes in visited(black): map[]

current node: node 0
visit descendant: node 1
nodes in visiting(gray): map[0:true 1:true]
nodes in visited(black): map[]

current node: node 1
visit descendant: node 2
nodes in visiting(gray): map[0:true 1:true 2:true]
nodes in visited(black): map[]

finish explore node 2
nodes in visiting(gray): map[0:true 1:true]
nodes in visited(black): map[2:true]

current node: node 1
visit descendant: node 3
nodes in visiting(gray): map[0:true 1:true 3:true]
nodes in visited(black): map[2:true]

current node: node 3
visit descendant: node 4
nodes in visiting(gray): map[0:true 1:true 3:true 4:true]
nodes in visited(black): map[2:true]

current node: node 4
visit descendant: node 5
nodes in visiting(gray): map[0:true 1:true 3:true 4:true 5:true]
nodes in visited(black): map[2:true]

current node: node 5
visit descendant: node 3
node 3 is in visiting(gray) -> a cycle is detected
*/

File: LeetCode 207- course schedule/course_schedule.py
class Solution(object):
    def canFinish(self, numCourses, prerequisites):
        """
        :type numCourses: int
        :type prerequisites: List[List[int]]
        :rtype: bool
        """
        graph = self.buildGraph(numCourses, prerequisites)
        print("adjacency list", graph)
        print("")
        visiting = set()
        visited = set()
        for node in range(numCourses):
            print("----------------------------------------------")
            print("dfs node", node)
            if self.detectCycle(graph, node, visited, visiting):
                return False
        
        return len(visited) == numCourses
            
    def detectCycle(self, graph, node, visited, visiting):
        if node in visited:
            print("node ", node, "is already visited(black) -> skip")
            return False
        
        if node in visiting:
            print("node ", node, "is in visiting(gray) -> a cycle is detected")
            print("")
            return True
        
        visiting.add(node)
        print("nodes in visiting(gray)", visiting)
        print("nodes in visited(black)", visited)
        print("")
        for descendant in graph[node]:
            print("current node:", node)
            print("visit descendant: node", descendant)
            if self.detectCycle(graph, descendant, visited, visiting):
                return True
            print("")
    
        visiting.remove(node)
        visited.add(node)
        print("finish explored node", node)
        print("nodes in visiting(gray)", visiting)
        print("nodes in visited(black)", visited)
        return False
        
        
    def buildGraph(self, numNodes, edges):
        graph = {x: [] for x in range(numNodes)}
        for edge in edges:
            a, b = edge
            graph[b].append(a)
        return graph
        

if __name__ == "__main__":
    solution = Solution()
    numCourses = 6
    prerequisites = [[1, 0], [3, 0], [5, 0], [2, 1], [3, 1], [4, 3], [5, 4], [3, 5]]
    print("can finish courses?", solution.canFinish(numCourses, prerequisites))

"""
output:
('adjacency list', {0: [1, 3, 5], 1: [2, 3], 2: [], 3: [4], 4: [5], 5: [3]})

----------------------------------------------
('dfs node', 0)
('nodes in visiting(gray)', set([0]))
('nodes in visited(black)', set([]))

('current node:', 0)
('visit descendant: node', 1)
('nodes in visiting(gray)', set([0, 1]))
('nodes in visited(black)', set([]))

('current node:', 1)
('visit descendant: node', 2)
('nodes in visiting(gray)', set([0, 1, 2]))
('nodes in visited(black)', set([]))

('finish explored node', 2)
('nodes in visiting(gray)', set([0, 1]))
('nodes in visited(black)', set([2]))

('current node:', 1)
('visit descendant: node', 3)
('nodes in visiting(gray)', set([0, 1, 3]))
('nodes in visited(black)', set([2]))

('current node:', 3)
('visit descendant: node', 4)
('nodes in visiting(gray)', set([0, 1, 3, 4]))
('nodes in visited(black)', set([2]))

('current node:', 4)
('visit descendant: node', 5)
('nodes in visiting(gray)', set([0, 1, 3, 4, 5]))
('nodes in visited(black)', set([2]))

('current node:', 5)
('visit descendant: node', 3)
('node ', 3, 'is in visiting(gray) -> a cycle is detected')

('can finish courses?', False)
"""